JURTKiller fails with java.lang.ClassNotFoundException: Illegal access using Tomcat 9

Seen when running on top of Tomcat 9.0.46 and reloading an app:
```
05-Jan-2022 10:21:02.632 INFO [JURTKiller] org.apache.catalina.loader.WebappClassLoaderBase.checkStateForResourceLoading Illegal access: this web application instance has been stopped already. Could not load [java.lang.Thread$State]. The following sta>
        java.lang.IllegalStateException: Illegal access: this web application instance has been stopped already. Could not load [java.lang.Thread$State]. The following stack trace is thrown for debugging purposes as well as to attempt to terminate the>
                at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForResourceLoading(WebappClassLoaderBase.java:1384)
                at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForClassLoading(WebappClassLoaderBase.java:1372)
                at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1225)
                at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1187)
                at se.jiderhamn.classloader.leak.prevention.cleanup.StopThreadsCleanUp$JURTKiller.run(StopThreadsCleanUp.java:347)
Exception in thread "JURTKiller" java.lang.NoClassDefFoundError: java/lang/Thread$State
        at se.jiderhamn.classloader.leak.prevention.cleanup.StopThreadsCleanUp$JURTKiller.run(StopThreadsCleanUp.java:347)
Caused by: java.lang.ClassNotFoundException: Illegal access: this web application instance has been stopped already. Could not load [java.lang.Thread$State]. The following stack trace is thrown for debugging purposes as well as to attempt to terminate>
        at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForClassLoading(WebappClassLoaderBase.java:1374)
        at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1225)
        at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1187)
        ... 1 more
Caused by: java.lang.IllegalStateException: Illegal access: this web application instance has been stopped already. Could not load [java.lang.Thread$State]. The following stack trace is thrown for debugging purposes as well as to attempt to terminate >
        at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForResourceLoading(WebappClassLoaderBase.java:1384)
        at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForClassLoading(WebappClassLoaderBase.java:1372)
        ... 3 more
```


It turns out that the `JURTKiller` thread, spawned to force the OpenOffice `AsynchronousFinalizer` to run and kill it after completion, is attempting to load the `java.lang.Thread$State` enum during its execution, 5 seconds after it was started.
Problem is that the underlying classloader has already been "locked" by Tomcat as the app has been unloaded, and prevents any class to be loaded from this point on.

As a result, the thread started by the OpenOffice `AsynchronousFinalizer` is never stopped and leaks the classloader, defeating the purpose of the `JURTKiller`implementation entirely.

--- a/classloader-leak-prevention/classloader-leak-prevention-core/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/StopThreadsCleanUp.java
+++ b/classloader-leak-prevention/classloader-leak-prevention-core/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/StopThreadsCleanUp.java
@@ -71,7 +71,7 @@ public class StopThreadsCleanUp implements ClassLoaderPreMortemCleanUp {
     stopThreads(preventor);
   }
   
-  /** 
+  /**
    * The bug detailed at https://issues.apache.org/ooo/show_bug.cgi?id=122517 is quite tricky. This is a try to 
    * avoid the issues by force starting the threads and it's job queue.
    */
@@ -310,11 +310,11 @@ public class StopThreadsCleanUp implements ClassLoaderPreMortemCleanUp {
    * See https://issues.apache.org/ooo/show_bug.cgi?id=122517
    */
   protected class JURTKiller extends Thread {
-    
+
     private final ClassLoaderLeakPreventor preventor;
-    
+
     private final Thread jurtThread;
-    
+
     private final List<?> jurtQueue;
 
     public JURTKiller(ClassLoaderLeakPreventor preventor, Thread jurtThread) {
@@ -322,46 +322,54 @@ public class StopThreadsCleanUp implements ClassLoaderPreMortemCleanUp {
       this.preventor = preventor;
       this.jurtThread = jurtThread;
       jurtQueue = preventor.getStaticFieldValue(StopThreadsCleanUp.JURT_ASYNCHRONOUS_FINALIZER, "queue");
+      // Make sure all classes are loaded from the current app classloader before it executes,
+      // as it may use them *after* the classloader has been "shutdown" by the container (if any).
+      State state = State.RUNNABLE;
     }
 
     @Override
     public void run() {
-      if(jurtQueue == null || jurtThread == null) {
-        preventor.error(getName() + ": No queue or thread!?");
-        return;
-      }
-      if(! jurtThread.isAlive()) {
-        preventor.warn(getName() + ": " + jurtThread.getName() + " is already dead?");
-      }
-      
-      boolean queueIsEmpty = false;
-      while(! queueIsEmpty) {
         try {
-          preventor.debug(getName() + " goes to sleep for " + ClassLoaderLeakPreventor.THREAD_WAIT_MS_DEFAULT + " ms");
-          Thread.sleep(ClassLoaderLeakPreventor.THREAD_WAIT_MS_DEFAULT);
-        }
-        catch (InterruptedException e) {
-          // Do nothing
-        }
+            if(jurtQueue == null || jurtThread == null) {
+                preventor.error(getName() + ": No queue or thread!?");
+                return;
+              }
+              if(! jurtThread.isAlive()) {
+                preventor.warn(getName() + ": " + jurtThread.getName() + " is already dead?");
+              }
 
-        if(State.RUNNABLE != jurtThread.getState()) { // Unless thread is currently executing a Job
-          preventor.debug(getName() + " about to force Garbage Collection");
-          ClassLoaderLeakPreventor.gc(); // Force garbage collection, which may put new items on queue
+              boolean queueIsEmpty = false;
+              while(! queueIsEmpty) {
+                try {
+                  preventor.debug(getName() + " goes to sleep for " + ClassLoaderLeakPreventor.THREAD_WAIT_MS_DEFAULT + " ms");
+                  Thread.sleep(ClassLoaderLeakPreventor.THREAD_WAIT_MS_DEFAULT);
+                }
+                catch (InterruptedException e) {
+                  // Do nothing
+                }
 
-          synchronized (jurtQueue) {
-            queueIsEmpty = jurtQueue.isEmpty();
-            preventor.debug(getName() + ": JURT queue is empty? " + queueIsEmpty);
-          }
+                if(State.RUNNABLE != jurtThread.getState()) { // Unless thread is currently executing a Job
+                  preventor.debug(getName() + " about to force Garbage Collection");
+                  ClassLoaderLeakPreventor.gc(); // Force garbage collection, which may put new items on queue
+
+                  synchronized (jurtQueue) {
+                    queueIsEmpty = jurtQueue.isEmpty();
+                    preventor.debug(getName() + ": JURT queue is empty? " + queueIsEmpty);
+                  }
+                }
+                else
+                  preventor.debug(getName() + ": JURT thread " + jurtThread.getName() + " is executing Job");
+                }
+
+              preventor.info(getName() + " about to kill " + jurtThread);
+              if(jurtThread.isAlive()) {
+                //noinspection deprecation
+                jurtThread.stop();
+              }
         }
-        else 
-          preventor.debug(getName() + ": JURT thread " + jurtThread.getName() + " is executing Job");
-      }
-      
-      preventor.info(getName() + " about to kill " + jurtThread);
-      if(jurtThread.isAlive()) {
-        //noinspection deprecation
-        jurtThread.stop();
-      }
+        catch (Throwable t) {
+            preventor.error(t);
+        }
     }
   }
   

